### DISCUSSION TITLE
Crash on aqua:gitlab.com/gitlab-org/cli install

### DISCUSSION THREAD

**Discussion by scop** [jdx/mise#3547] [Troubleshooting and bug reports]
**Describe the bug**

mise crashes on installing aqua:gitlab.com/gitlab-org/cli.

**To Reproduce**

`mise use -g aqua:gitlab.com/gitlab-org/cli`

**Expected behavior**

No crash, `glab` installed.

**`mise doctor` output**

```text
version: 2024.11.37 linux-x64 (7ac5ab3 2024-11-30)
activated: yes
shims_on_path: no

build_info:
  Target: x86_64-unknown-linux-gnu
  Features: DEFAULT, NATIVE_TLS, OPENSSL
  Built: Sat, 30 Nov 2024 01:03:00 +0000
  Rust Version: rustc 1.82.0 (f6e511eec 2024-10-15)
  Profile: release

shell:
  /bin/bash
  GNU bash, version 5.2.21(1)-release (x86_64-pc-linux-gnu)
  Copyright (C) 2022 Free Software Foundation, Inc.
  License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

  This is free software; you are free to change and redistribute it.
  There is NO WARRANTY, to the extent permitted by law.

dirs:
  cache: ~/.cache/mise
  config: ~/.config/mise
  data: ~/.local/share/mise
  shims: ~/.local/share/mise/shims
  state: ~/.local/state/mise

config_files:
  ~/.config/mise/config.toml

backends:
  aqua
  asdf
  cargo
  core
  gem
  go
  npm
  pipx
  spm
  ubi
  vfox

plugins:
  dprint  https://github.com/asdf-community/asdf-dprint#79889ec
  usage   https://github.com/jdx/mise-usage.git#fe3888a

toolset:
  aqua:astral-sh/uv@0.5.5
  aqua:cli/cli@2.63.0
  aqua:crate-ci/typos@1.28.1
  aqua:evilmartians/lefthook@1.8.4
  aqua:goreleaser/nfpm@2.41.1
  aqua:gotestyourself/gotestsum@1.12.0
  aqua:mvdan/sh@3.10.0
  asdf:dprint@0.47.5
  asdf:usage@1.3.3
  core:go@1.23.3
  ubi:aquaproj/aqua@2.38.1
  ubi:golangci/golangci-lint@1.62.2
  ubi:goreleaser/goreleaser@2.4.8
  ubi:koalaman/shellcheck@0.10.0
  ubi:suzuki-shunsuke/cmdx@1.7.4

env_vars:
  MISE_USE_TOML=true
  MISE_SHELL=bash

settings:

1 warning found:

1. plugin python overrides a core plugin

No problems found
```

**Additional context**

```shellsession
$ RUST_BACKTRACE=1 mise --debug use -g "aqua:gitlab.com/gitlab-org/cli"
DEBUG ARGS: mise --debug use -g aqua:gitlab.com/gitlab-org/cli
DEBUG config: ~/.config/mise/config.toml
DEBUG install_some_versions: aqua:gitlab.com/gitlab-org/cli@latest
INFO  aqua:gitlab.com/gitlab-org/cli@latest install
The application panicked (crashed).
Message:  called `Result::unwrap()` on an `Err` value: reqwest::Error { kind: Builder, source: RelativeUrlWithoutBase }
Location: src/http.rs:83

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 8 frames hidden ⋮                               
   9: mise::http::Client::head::hf337410e89320aac
      at <unknown source file>:<unknown line>
  10: mise::backend::aqua::AquaBackend::fetch_url::hfe48d6c1d73dc83c
      at <unknown source file>:<unknown line>
  11: <mise::backend::aqua::AquaBackend as mise::backend::Backend>::install_version_::h2c271abff9df44aa
      at <unknown source file>:<unknown line>
  12: mise::backend::Backend::install_version::h9a6449a4599c50c9
      at <unknown source file>:<unknown line>
  13: std::sys::backtrace::__rust_begin_short_backtrace::h1a6635786c8f2a07
      at <unknown source file>:<unknown line>
  14: core::ops::function::FnOnce::call_once{{vtable.shim}}::hd4f76bfb90e81c2b
      at <unknown source file>:<unknown line>
  15: std::sys::pal::unix::thread::Thread::new::thread_start::hcdbd1049068002f4
      at <unknown source file>:<unknown line>
  16: start_thread<unknown>
      at ./nptl/pthread_create.c:447
  17: clone3<unknown>
      at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
```

**Comment by jdx** [comment]
perhaps the docs need to be updated but I think you need to specify `https://` it's possible github.com is an exception, we could add gitlab too if that's the case.

**Comment by scop** [comment]
Hmm, not sure how to specify https there.
```
$ mise use -g aqua:https://gitlab.com/gitlab-org/cli
mise ERROR HTTP status client error (404 Not Found) for url (https://mise-versions.jdx.dev/aqua-registry/https://gitlab.com/gitlab-org/cli/registry.yaml)
mise ERROR Run with --verbose or MISE_VERBOSE=1 for more information
```

The aqua registry does have explicit gitlab.com/gitlab-org/cli path for this, I haven't checked if there are any others with the domain in it:
- https://github.com/aquaproj/aqua-registry/blob/102040dc4772620b882771d81f7ecb258a6fd3fb/registry.yaml#L23608
- https://github.com/aquaproj/aqua-registry/tree/8bc0821b056b7c1c196557ea2ad295c3a9ebb0b2/pkgs/gitlab.com/gitlab-org/cli

Anyway I thought the crash was the more interesting thing here rather than failure to install (be it due to using a wrong syntax or not).

**Comment by jdx** [comment]
oh sorry, I forgot, everything in aqua is hardcoded to github. I hadn't seen anything in there that wasn't using github. In fact it seems this is something handled as a special case because the registry is specified as http: https://github.com/aquaproj/aqua-registry/blob/main/pkgs/gitlab.com/gitlab-org/cli/registry.yaml#L3-L5

I don't think we can support this one unless we also add in similar exceptions

**Comment by jdx** [comment]
I think I'm inclined to wait on this one until aqua more formally defines gitlab, other I'd basically need to handle this one package uniquely in the repo in many ways

**Comment by scop** [comment]
That's all fine with me, thanks for considering. Although perhaps the crash could be fixed meanwhile, even if installing the thing would not work.

> I hadn't seen anything in there that wasn't using github.

Just thought I'd note that at least 1password is specified as http in aqua too, but it's not a GitHub thing, and installing it via mise works fine (albeit without automatic remote version info).
```
$ mise install aqua:1password/cli@2.30.3
mise aqua:1password/cli@2.30.3 ✓ installed                                       
 ```

**Comment by jdx** [comment]
http is not a problem, using gitlab is since this appears to be the only tool in aqua not using github

**Comment by scop** [comment]
The mentioned `1password/cli` in aqua is not using github. Neither is `slack.com/slack-cli`, and both of them install fine.

**Comment by jdx** [comment]
we don't use either one, and it doesn't appear mise can get the versions for either one

**Comment by scop** [comment]
I'm not sure what "we don't use" means. `aqua:1password/cli` exists in mise's registry. `aqua:slack.com/slack.cli` does not, but that in general does not mean one cannot use mise to install it (or just about any arbitrary aqua registry entry even if it's not in mise's registry). Or is that ability a coincidence?

Anyway, https://mise.jdx.dev/registry.html even explicitly says
> You can also specify the full name for a tool using `mise use aqua:1password/cli` if you want to use a specific backend.

So, to highlight the difference and that it does not seem to be about GitHubness:

- `mise install aqua:1password/cli@v2.30.3` works (you're right mise can't get versions for this so `@latest` does not work, but that's unrelated)
- `mise install aqua:slack.com/slack-cli` works, although that can be a coincidence as there happens to be a `latest` version available

On the other hand:

- `mise install aqua:gitlab.com/gitlab-org/cli` crashes, as mentioned
- `mise install aqua:gitlab.com/gitlab-org/cli@v1.50.0` crashes as well

**Comment by jdx** [comment]
Tools only "work" in the strictest sense in mise if we can't list the versions. So much of the functionality doesn't work if we can't list versions that I don't really consider something "working" if we can't list them. That's why 1password doesn't use aqua by default.

I take your point that something other than listing versions is making this tool not work, but unless we were able to list the versions I wouldn't consider putting it into the registry anyways.

**Comment by scop** [comment]
I agree with version listing availability being a good "gatekeeper" for what gets included in mise's registry.

But I hope the general ability to just use something from aqua (or ubi for that matter) no matter if they're in mise's registry or not does not go away or become a coincidence. In some use cases, version listing is entirely moot, and tools are always installed or exec'd by explicit versions, and mise's ability to transparently download them if needed on the fly is very useful.

Anyway, we digress -- this bug report is not about version support or whether a tool is supported for one reason or the other, but that in my (not so humble :)) opinion, _crashing_ on something like the reproducer here is not appropriate. Gracefully exiting with a non-zero exit code and a descriptive message would be fine.

**Comment by jdx** [comment]
from a practical standpoint, this is the only tool in aqua that uses gitlab. It appears from the aqua registry given the TODO there are plans to more formally support gitlab. I think at that point, once aqua has a `gitlab_release` type or something it would be worth fixing this. I just don't think it's worth spending any of my time on this—including changing the error output—until then because it only affects one tool and I'll eventually need to revisit it when aqua gitlab support lands.

**Comment by jdx** [comment]
I don't have an example but I believe there are `http` packages in aqua that do support listing versions without relying on github releases/tags somehow so the fix for 1password and slack is likely for that information to go in the aqua registry

**Comment by Marukome0743** [comment]
The crash reported here is now fixed by #11160 (merged, will ship in the next release).

Several `url.into_url().unwrap()` sites in the HTTP layer previously panicked the whole process with `RelativeUrlWithoutBase` (the `src/http.rs:83` crash in the backtrace above) whenever a backend handed a relative or otherwise invalid URL to the client. Those now return a graceful error instead of aborting — i.e. the "gracefully exiting with a non-zero exit code and a descriptive message" behavior @scop asked for.

To be clear about scope: this fixes the **panic class**, not aqua's GitLab support. Getting `glab` itself installed is best done through the native `gitlab:` backend today (`mise use gitlab:gitlab-org/cli`); the "aqua only hardcodes github" discussion is still tracked separately and depends on aqua formally defining a gitlab release type, as @jdx noted.

*This comment was generated by an AI coding assistant.*
